src/lib/transition_frontier/transition_frontier.ml

```diff
@@ -1 +1,275 @@
+    val apply : t -> Ledger_builder_diff.t -> unit
+  end
+
+  module Merkle_ledger : sig
+    type t
+
+    val root : t -> Ledger_hash.t
+    val derive : t -> Merkle_mask.t
+    val merge : t -> Merkle_mask.t -> unit
+  end
+
+  module Max_length : sig
+    val t : int
+  end
+end
+
+(* NOTE: is Consensus_mechanism.select preferable over distance? *)
+module Make
+    (Inputs : Inputs_intf)
+  : Transition_frontier_intf
+    with type state_hash := Inputs.State_hash.t
+     and type external_transition := Inputs.External_transition.t
+     and type merkle_ledger := Inputs.Merkle_ledger.t
+     and type merkle_mask := Inputs.Merkle_mask.t =
+struct
+  open Inputs
+
+  let max_length = Max_length.t
+
+  module Breadcrumb = struct
+    type t =
+      { transition: External_transition.t
+      ; mask: Merkle_mask.t }
+    [@@deriving fields]
+
+    let hash t =
+      t.transition
+      |> External_transition.protocol_state
+      |> Protocol_state.hash
+
+    let parent_hash t =
+      t.transition
+      |> External_transition.protocol_state
+      |> Protocol_state.previous_state_hash
+  end
+
+  type node =
+    { breadcrumb: Breadcrumb.t
+    ; successor_hashes: State_hash.t list
+    ; length: int }
+
+  type t =
+    { root_ledger: Merkle_ledger.t
+    ; mutable root: State_hash.t
+    ; mutable best_tip: State_hash.t
+    ; table: node State_hash.Table.t }
+
+  (* TODO: load from and write to disk *)
+  let create ~root ~ledger =
+    let protocol_state = External_transition.protocol_state root in
+    let blockchain_state = Protocol_state.blockchain_state protocol_state in
+    assert (
+      Ledger_hash.equal
+        (Merkle_ledger.root ledger)
+        (Frozen_ledger_hash.to_ledger_hash
+           (Blockchain_state.ledger_hash blockchain_state)));
+
+    let mask = Merkle_ledger.derive ledger in
+    Merkle_mask.apply mask (External_transition.ledger_builder_diff root);
+    assert (
+      Ledger_hash.equal
+        (Merkle_mask.root mask)
+        (Ledger_builder_hash.ledger_hash
+           (Blockchain_state.ledger_builder_hash blockchain_state)));
+
+    let root_hash = Protocol_state.hash protocol_state in
+    let root_breadcrumb = {Breadcrumb.transition=root; mask} in
+    let root_node = {breadcrumb=root_breadcrumb; successor_hashes=[]; length=0} in
+    let table = State_hash.Table.of_alist_exn [(root_hash, root_node)] in
+    { root_ledger= ledger
+    ; root= root_hash
+    ; best_tip= root_hash
+    ; table }
+
+  let find t hash =
+    let open Option.Let_syntax in
+    let%map node = Hashtbl.find t.table hash in
+    node.breadcrumb
+
+  let find_exn t hash =
+    let node = Hashtbl.find_exn t.table hash in
+    node.breadcrumb
+
+  let path t breadcrumb =
+    let rec find_path b =
+      let hash = Breadcrumb.hash b in
+      let parent_hash = Breadcrumb.parent_hash b in
+      if State_hash.equal parent_hash t.root then
+        [hash]
+      else
+        hash :: find_path (find_exn t parent_hash)
```

Comment:
Somewhere we should validate the correctness of breadcrumbs we get from the network such that we can be sure that `find_exn` is safe. Any ideas where?

Reply:
Discussed in person.